cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,502 @@
+ *       in B and the thread y is the starting column in B
+ * 
+ * @param[in]   A    Pointer to the raw data of matrix `A`
+ * @param[in]   m    Number of rows (A)
+ * @param[in]   n    Number of columns (A)
+ * @param[in]   B    Pointer to the raw data of matrix `B`
+ * @param[in]   p    Number of rows (B)
+ * @param[in]   q    Number of columns (B)
+ * @param[out]  AkB  Pointer to raw data of the result kronecker product
+ * @param[in]   k_m  Number of rows of the result    (m * p)
+ * @param[in]   k_n  Number of columns of the result (n * q)
+ */
+__global__ void kronecker_product_kernel(const double* A, int m, int n,
+                                         const double* B, int p, int q,
+                                         double* AkB, int k_m, int k_n) {
+  const double* A_b = A + blockIdx.x * m * n;
+  const double* B_b = B + blockIdx.x * p * q;
+  double* AkB_b = AkB + blockIdx.x * k_m * k_n;
+
+  for (int ia = 0; ia < m; ia++) {
+    for (int ja = 0; ja < n; ja++) {
+      double A_ia_ja = A_b[ia + ja * m];
+
+      for (int ib = threadIdx.x; ib < p; ib += blockDim.x) {
+        for (int jb = threadIdx.y; jb < q; jb += blockDim.y) {
+          int i_ab = ia * p + ib;
+          int j_ab = ja * q + jb;
+          AkB_b[i_ab + j_ab * k_m] = A_ia_ja * B_b[ib + jb * p];
+        }
+      }
+    }
+  }
+}
+
+/**
+ * @brief Multiplies each matrix in a batch-A with it's batch-B counterpart.
+ *        A = [A1,A2,A3], B=[B1,B2,B3] returns [A1*B1, A2*B2, A3*B3]
+ * 
+ * @param[in]  A   First matrix batch
+ * @param[in]  B   Second matrix batch
+ * @param[in]  aT  Is `A` transposed?
+ * @param[in]  bT  is `B` transposed?
+ * 
+ * @return Member-wise A*B
+ */
+BatchedMatrix b_gemm(const BatchedMatrix& A, const BatchedMatrix& B,
+                     bool aT = false, bool bT = false) {
+  if (A.batches() != B.batches()) {
+    throw std::runtime_error("A & B must have same number of batches");
+  }
+
+  // Logic for matrix dimensions with optional transpose
+  // m = number of rows of matrix op(A) and C.
+  int m = !aT ? A.shape().first : A.shape().second;
+
+  // n = number of columns of matrix op(B) and C.
+  int n = !bT ? B.shape().second : B.shape().first;
+
+  // k = number of columns of op(A) and rows of op(B).
+  int k = !aT ? A.shape().second : A.shape().first;
+  int kB = !bT ? B.shape().first : B.shape().second;
+
+  if (k != kB) {
+    throw std::runtime_error("Matrix-Multiplication dimensions don't match!");
+  }
+
+  auto num_batches = A.batches();
+  const auto& handle = A.cublasHandle();
+
+  // set transpose
+  cublasOperation_t opA = aT ? CUBLAS_OP_T : CUBLAS_OP_N;
+  cublasOperation_t opB = bT ? CUBLAS_OP_T : CUBLAS_OP_N;
+
+  // Create C(m,n)
+  BatchedMatrix C(m, n, num_batches, A.cublasHandle(), A.allocator(),
+                  A.stream());
+
+  double alpha = 1.0;
+  double beta = 0.0;
+
+  // [C1,C2,C3] = [A1*B1, A2*B2, A3*B3]
+  CUBLAS_CHECK(
+    MLCommon::LinAlg::cublasgemmBatched(handle,
```

Comment:
I believe the only reason we are allocating pointer-to-pointers is because of this batched-gemm call?
If so, @Nyrio and/or @rietmann-nv, why not we use `cublas?gemmStridedBatched` to avoid this altogether?

Reply:
As dicussed, I was able to remove `double**` for matrix-matrix multiplications, but I realized that we still need it for the matrix solve.

Reply:
Got it. Fair enough! Let's keep the `double**` case just for matrix solve then. Thanks.